src/jit/lower.cpp

```diff
@@ -2976,6 +2979,562 @@ GenTree* Lowering::LowerCompare(GenTree* cmp)
+        {
+            node = node->gtNext;
+        }
+
+        return node;
+    }
+
+    static GenTree* PreviousInstruction(GenTree* node)
+    {
+        node = node->gtPrev;
+
+        while ((node != nullptr) && node->OperIs(GT_IL_OFFSET))
+        {
+            node = node->gtPrev;
+        }
+
+        return node;
+    }
+
+    static GenTree* LastInstruction(BasicBlock* block)
+    {
+        GenTree* last = block->lastNode();
+
+        while ((last != nullptr) && last->OperIs(GT_IL_OFFSET))
+        {
+            last = last->gtPrev;
+        }
+
+        return last;
+    }
+
+    static BasicBlock* SingleSuccessorBlock(BasicBlock* block)
+    {
+        if (block->bbJumpKind == BBJ_NONE)
+        {
+            return block->bbNext;
+        }
+
+        if (block->bbJumpKind == BBJ_ALWAYS)
+        {
+            return block->bbJumpDest;
+        }
+
+        return nullptr;
+    }
+
+    static GenTree* CloneNode(Compiler* comp, GenTree* node)
+    {
+        if (node->OperIs(GT_STORE_LCL_VAR))
+        {
+            GenTreeLclVar* clone = new (comp, GT_STORE_LCL_VAR)
+                GenTreeLclVar(GT_STORE_LCL_VAR, node->TypeGet(), node->AsLclVar()->GetLclNum(), -1);
+            clone->gtFlags = node->gtFlags & GTF_LIVENESS_MASK;
+            comp->lvaIncRefCnts(clone);
+            return clone;
+        }
+        else if (node->OperIs(GT_LCL_VAR))
+        {
+            GenTree* clone = comp->gtClone(node);
+            comp->lvaIncRefCnts(clone);
+            return clone;
+        }
+        else if (node->OperIs(GT_RETURN))
+        {
+            return comp->gtNewOperNode(GT_RETURN, node->TypeGet(), node->gtGetOp1());
+        }
+        else
+        {
+            assert(node->OperIs(GT_CNS_INT));
+            return comp->gtClone(node);
+        }
+    }
+
+    void UnlinkBlocks(BasicBlock* falseBlock) const
+    {
+        assert(m_joinBlock != nullptr);
+
+        m_comp->fgRemoveRefPred(falseBlock, m_block);
+
+        m_block->bbJumpKind = BBJ_ALWAYS;
+        m_block->bbJumpDest = m_joinBlock;
+
+        m_comp->fgAddRefPred(m_joinBlock, m_block);
+    }
+
+    void UnlinkBlocks(BasicBlock* falseBlock, BasicBlock* trueBlock) const
+    {
+        m_comp->fgRemoveRefPred(falseBlock, m_block);
+        m_comp->fgRemoveRefPred(trueBlock, m_block);
+
+        if (m_joinBlock == nullptr)
+        {
+            m_block->bbJumpKind = BBJ_RETURN;
+        }
+        else
+        {
+            m_block->bbJumpKind = BBJ_ALWAYS;
+            m_block->bbJumpDest = m_joinBlock;
+            m_joinBlock->bbFlags |= BBF_JMP_TARGET;
+            m_comp->fgAddRefPred(m_joinBlock, m_block);
+        }
+    }
+
+    LIR::Range& BlockRange() const
+    {
+        return LIR::AsRange(m_block);
+    }
+
+    enum class HammockKind
+    {
+        None,
+        Half,
+        Full
+    };
+
+    HammockKind MakeHammock()
+    {
+        m_falseBlock = m_block->bbNext;
+        m_trueBlock  = m_block->bbJumpDest;
+
```

Comment:
Also make sure that neither of these is `block` itself.

Reply:
I sometimes try and forget that the jit's flowgraph has implicit fall through semantics. 

But doing an early check that `block` and `m_trueBlock` are different may save a bit of time.

Reply:
The interesting thing is that if `trueBlock == block` you can still have a correct half hammock if `falseBlock` jumps back to `block`. But then this would be an infinite loop which we don't care about (and likely don't have test coverage for).

And I should take a look at other scenarios involving loops (e.g. `joinBlock == block`).